### PR TITLE
feat: docker l2tp driver

### PR DESCRIPTION
**Pull request by oopcode**
This PR makes it possible for containers to connect to L2TP networks (no IPsec) via Docker's IPAM/Network drivers.
Each container gets its own Docker network; the PPP interface that's created by `xl2tpd` is moved into container's namespace, the container can reach private network via this interface.
N.B.: `xl2tpd` should be running on Worker (host).

### CODE REVIEW

File: insonmnia/miner/network/l2tp_store.go [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+
+func newL2tpNetwork(opts *L2TPNetworkConfig) *l2tpNetwork {
+	return &l2tpNetwork{
+		NetworkOpts: opts,
+		Store:       NewL2TPEndpointStore(),
+	}
+}
+
+func (n *l2tpNetwork) Setup() error {
+	n.PoolID = n.NetworkOpts.GetHash()
```
Comment (antmat) [inline insonmnia/miner/network/l2tp_store.go]: why does pool id is bound to options? What if I want to run 2 containers on the same host, connected to the same l2tp server?

File: insonmnia/miner/network/l2tp_tuner.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	netListener, err := net.Listen("unix", t.cfg.NetSocketPath)
+	if err != nil {
+		log.G(context.Background()).Error("Failed to listen", zap.Error(err))
+		return err
+	}
+
+	netHandler := netDriver.NewHandler(t.netDriver)
+
+	syscall.Unlink(t.cfg.IPAMSocketPath)
+	ipamListener, err := net.Listen("unix", t.cfg.IPAMSocketPath)
```
Comment (antmat) [inline insonmnia/miner/network/l2tp_tuner.go]: Maybe we can generalize plugin startup somehow?

File: insonmnia/miner/network/l2tp_tuner.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+func (t *L2TPTuner) writeConfig(netID string, opts map[string]string) (string, error) {
+	var data string
+	for k, v := range opts {
+		data += fmt.Sprintf("%s: %s\n", k, v)
+	}
+
+	path := t.cfg.ConfigDir + "/" + netID
+
+	return path, ioutil.WriteFile(path, []byte(data), 700)
```
Comment (antmat) [inline insonmnia/miner/network/l2tp_tuner.go]: s/700/0700/

File: task.yaml [hunk 4]
```diff
@@ -10,32 +10,23 @@ task:
...
-#          share: samba-host.ru/share
-#          username: username
-#          password: password
-#          security: ntlm
-#          vers: 3.0
-#    mounts:
-#      - cifs:/mnt:rw
-#      - cifs:/opt:rw
-  # container resources definition, required section
-  resources:
```
[the remark this answers is not in the document]
Comment (oopcode) [reply task.yaml]: That was a request from  @sshaman1101.

File: insonmnia/miner/plugin/config.go [hunk 5]
```diff
@@ -3,10 +3,11 @@ package plugin
...
 
 type Config struct {
-	SocketDir string        `yaml:"socket_dir" default:"/run/docker/plugins"`
-	Volumes   VolumesConfig `yaml:"volume"`
-	GPUs      map[string]map[string]string
+	SocketDir string                     `yaml:"socket_dir" default:"/run/docker/plugins"`
+	Volumes   VolumesConfig              `yaml:"volume"`
 	Tinc      *network.TincNetworkConfig `yaml:"tinc"`
+	L2TP      *network.L2TPConfig        `yaml:"l2tp"`
+	GPUs      map[string]map[string]string
```
[the remark this answers is not in the document]
Comment (oopcode) [reply insonmnia/miner/plugin/config.go]: this keeps field tags aligned properly

File: insonmnia/miner/network/l2tp_config.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+	PPPDefaultRoute     bool   `required:"false" yaml:"ppp_default_route" default:"true"`
+	PPPUsepeerdns       bool   `required:"false" yaml:"ppp_use_peer_dns" default:"true"`
+	PPPLock             bool   `required:"false" yaml:"ppp_lock" default:"true"`
+	PPPIPCPAcceptLocal  bool   `required:"false" yaml:"ppp_ipcp_accept_local" default:"true"`
+	PPPIPCPAcceptRemote bool   `required:"false" yaml:"ppp_ipcp_accept_remote" default:"true"`
+	PPPRefuseEAP        bool   `required:"false" yaml:"ppp_refuse_eap" default:"true"`
+	PPPRequireMSChapV2  bool   `required:"false" yaml:"ppp_require_mschap_v2" default:"true"`
+}
+
+func (o *l2tpNetworkConfig) GetPoolID() string {
```
Comment (3Hren) [inline insonmnia/miner/network/l2tp_config.go]: Just `PoolID()`.

File: insonmnia/miner/network/l2tp_config.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+	if !ok {
+		return nil, errors.New("config path not provided")
+	}
+
+	cfg := &l2tpNetworkConfig{}
+	if err := configor.Load(cfg, path.(string)); err != nil {
+		return nil, err
+	}
+
+	if err := cfg.validate(); err != nil {
```
Comment (3Hren) [inline insonmnia/miner/network/l2tp_config.go]: Just `return cfg.validate()` but whatever.

File: insonmnia/miner/network/l2tp_ipam.go [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+			zap.String("network_id", l2tpNet.ID), zap.Error(err))
+		return err
+	}
+
+	return nil
+}
+
+func (d *IPAMDriver) ReleaseAddress(request *ipam.ReleaseAddressRequest) error {
+	d.logger.Infow("received ReleaseAddress request", zap.Any("request", request))
+	d.mu.Lock()
```
Comment (3Hren) [inline insonmnia/miner/network/l2tp_ipam.go]: Lock to just to unlock?

File: insonmnia/miner/network/l2tp_network.go [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+			"l2tpd_conn_name": l2tpNet.Endpoint.ConnName,
+			"device_name":     l2tpNet.Endpoint.PPPDevName,
+		},
+	}, nil
+}
+
+func (d *L2TPNeworktDriver) AllocateNetwork(request *network.AllocateNetworkRequest) (*network.AllocateNetworkResponse, error) {
+	d.logger.Infow("received AllocateNetwork request", zap.Any("request", request))
+	d.mu.Lock()
+	defer d.mu.Unlock()
```
Comment (3Hren) [inline insonmnia/miner/network/l2tp_network.go]: Either incomplete or looks strange.

File: insonmnia/miner/network/l2tp_network.go [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+	defer d.state.Sync()
+
+	return nil, nil
+}
+
+func (d *L2TPNeworktDriver) DeleteNetwork(request *network.DeleteNetworkRequest) error {
```
Comment (3Hren) [inline insonmnia/miner/network/l2tp_network.go]: Either incomplete or looks strange.

File: insonmnia/miner/network/l2tp_network.go [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+	d.logger.Infow("received DeleteNetwork request", zap.Any("request", request))
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	defer d.state.Sync()
+
+	return nil
+}
+
+func (d *L2TPNeworktDriver) FreeNetwork(request *network.FreeNetworkRequest) error {
```
Comment (3Hren) [inline insonmnia/miner/network/l2tp_network.go]: Here and below: either incomplete or looks strange.

File: insonmnia/miner/network/l2tp_state.go [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+
+	if err := state.load(); err != nil {
+		return nil, err
+	}
+
+	return state, nil
+}
+
+func (s *l2tpState) Sync() (err error) {
+	defer func() {
```
Comment (3Hren) [inline insonmnia/miner/network/l2tp_state.go]: I'd rather instead make `sync() error` with the logic inside and `Sync` with error checking and logging.

File: insonmnia/miner/network/l2tp_state.go [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+func NewL2TPEndpoint(netInfo *l2tpNetwork) *l2tpEndpoint {
+	return &l2tpEndpoint{
+		Name:        fmt.Sprintf("%d_%s", netInfo.Count, netInfo.PoolID),
+		NetworkOpts: netInfo.NetworkOpts,
+	}
+}
+
+func (e *l2tpEndpoint) setup() error {
+	e.ConnName = e.Name + "-connection"
+	e.PPPDevName = ("ppp" + e.Name)[:15]
```
Comment (3Hren) [inline insonmnia/miner/network/l2tp_state.go]: Are you sure that the name will contain at least 12-bytes?

File: insonmnia/structs/resources.go [hunk 14]
```diff
@@ -109,10 +109,6 @@ type TaskResources struct {
 }
 
 func NewTaskResources(r *pb.TaskResourceRequirements) (*TaskResources, error) {
-	if r == nil {
```
Comment (3Hren) [inline insonmnia/structs/resources.go]: Out of the scope, man.

File: task.yaml [hunk 15]
```diff
@@ -22,20 +22,11 @@ task:
 #    mounts:
 #      - cifs:/mnt:rw
 #      - cifs:/opt:rw
-  # container resources definition, required section
```
Comment (3Hren) [inline task.yaml]: Why?

File: worker.yaml [hunk 16]
```diff
@@ -64,6 +64,8 @@ plugins:
     volumes:
 #      cifs: {}
 
-  network:
-    tinc:
-      enabled: true
+  tinc:
+    enabled: true
+  l2tp:
+    enabled: true
```
Comment (3Hren) [inline worker.yaml]: Looks strange. I'd rather make `l2tp` optional field. For example `l2tp: {}` configures the driver, otherwise skips.

**Comment by 3Hren** [comment]
Add the description how this thing works.

File: insonmnia/miner/network/l2tp_state.go (hunk 13 above)
Comment (oopcode) [reply insonmnia/miner/network/l2tp_state.go]: absolutely, it's based on an md5 hash

File: worker.yaml (hunk 16 above)
Comment (oopcode) [reply worker.yaml]: i have to talk to my manager

**Pull request merged**